### PR TITLE
Allow custom axis scales through common component

### PR DESCRIPTION
**Pull request by RoryStokes**

### CODE REVIEW

File: src/axis.rs [hunk 1]
```diff
@@ -0,0 +1,15 @@
+use std::rc::Rc;
+
+// Axis scaled value, expected to be between 0 and 1
+// except in the case where the value is outside of the axis range
+pub struct NormalisedValue(pub f32);
```
Comment (huntc) [inline src/axis.rs:3]: Can we please make this Rustdoc i.e. `///`?

File: src/axis.rs [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+
+pub trait AxisScale {
```
Comment (huntc) [inline src/axis.rs:6]: Rustdoc for public types please.

File: src/horizontal_axis.rs [hunk 3]
```diff
@@ -70,33 +84,23 @@ impl Component for HorizontalAxis {
...
-                        y + p.tick_len
-                    };
-                    let x = (p.x1 as f32 + (i as f32 + p.scale.start) * scale) as u32;
+                { for(p.scale.ticks().iter()).map(|AxisTick { location: NormalisedValue(normalised_location), label }| {
+                    let x = p.x1 as f32 + normalised_location * scale;
                     html! {
                     <>
                         <line x1={x.to_string()} y1={y.to_string()} x2={x.to_string()} y2={to_y.to_string()} class="tick" />
-                        <text x={(x + 1).to_string()} y={to_y.to_string()} text-anchor="start" class="text">{i}</text>
+                        <text x={(x + 1 as f32).to_string()} y={to_y.to_string()} text-anchor="start" transform-origin={format!("{} {}", x, to_y + 1)} class="text">{label.to_string()}</text>
```
Comment (huntc) [inline src/horizontal_axis.rs]: `+ 1.0` is also valid here.

File: src/time_axis_scale.rs [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+            time_to,
+            step,
+            scale,
+            label_format: label_format.unwrap_or(DEFAULT_LABEL_FORMAT.to_string()),
+        }
+    }
+}
+
+impl AxisScale for TimeAxisScale {
+    fn ticks(&self) -> Rc<Vec<AxisTick>> {
```
Comment (huntc) [inline src/time_axis_scale.rs]: I think we should return just a Vec at this point... not an Rc. We shouldn't assume the type of box that the user will use.

**Review by huntc — CHANGES_REQUESTED** [review]

**Comment by huntc** [comment]
Many clippy errors which are failing CI also. Don't forget `cargo clippy` is your friend. :-)

File: src/axis.rs [hunk 5]
```diff
@@ -1,14 +1,17 @@
...
+/// A trait for creating an axis scale. It has two functions:
+/// `ticks` returns a vector of axis ticks
+/// `normalise` normalises the input `value`
 pub trait AxisScale {
-    fn ticks(&self) -> Rc<Vec<AxisTick>>;
+    fn ticks(&self) -> Vec<AxisTick>;
     fn normalise(&self, value: f32) -> NormalisedValue;
 }
 
+/// An axis tick
```
Comment (huntc) [inline src/axis.rs]: Full stop?

File: src/horizontal_axis.rs [hunk 6]
```diff
@@ -50,7 +50,10 @@ impl PartialEq for Props {
             && self.y1 == other.y1
             && self.tick_len == other.tick_len
             && self.title == other.title
-            && Rc::ptr_eq(&self.scale, &other.scale)
+            && std::ptr::eq(
+                &*self.scale as *const dyn AxisScale as *const u8,
+                &*other.scale as *const dyn AxisScale as *const u8,
+            )
```
Comment (huntc) [inline src/horizontal_axis.rs:58]: Can we please document this in reference to https://github.com/rust-lang/rust/issues/46139? It seems like a valid comparison to me as we are comparing allocations.

File: src/horizontal_series.rs [hunk 7]
```diff
@@ -69,13 +69,19 @@ impl PartialEq for Props {
...
             && self.height == other.height
             && self.width == other.width
+            && std::ptr::eq(
+                &*self.horizontal_scale as *const dyn AxisScale as *const u8,
+                &*other.horizontal_scale as *const dyn AxisScale as *const u8,
+            )
+            && std::ptr::eq(
+                &*self.vertical_scale as *const dyn AxisScale as *const u8,
+                &*other.vertical_scale as *const dyn AxisScale as *const u8,
+            )
```
Comment (huntc) [inline src/horizontal_series.rs:86]: May be document both of these referencing https://github.com/rust-lang/rust/pull/80505 too.

File: src/horizontal_axis.rs (hunk 6 above)
Comment (huntc) [reply src/horizontal_axis.rs:58]: Actually, can be written:

```rust
&*self.scale as *const _ *const u8
```

**Review by huntc — CHANGES_REQUESTED** [review]
Nearly there!

File: src/axis.rs (hunk 5 above)
Comment (RoryStokes) [reply src/axis.rs]: Expanded on the documentation including justifications

**Review by huntc — APPROVED** [review]
Thank you!

**Pull request merged**